pkg/maps/ctmap/per_cluster_ctmap.go

```diff
@@ -102,379 +123,124 @@ type PerClusterCTMapVal struct {
-				}
-			}
-		}
-	}()
+// NewPerClusterCTMaps returns a new instance of the per-cluster CT maps manager.
+func NewPerClusterCTMaps(ipv4, ipv6 bool) *perClusterCTMaps {
+	gm := perClusterCTMaps{clusterIDs: sets.New[uint32]()}
 
 	if ipv4 {
-		gm.tcp4, err = newPerClusterCTMap(outerMapNamePrefix+perClusterTCP4OuterMapSuffix, mapTypeIPv4TCPGlobal)
-		if err != nil {
-			return nil, err
-		}
-
-		gm.any4, err = newPerClusterCTMap(outerMapNamePrefix+perClusterANY4OuterMapSuffix, mapTypeIPv4AnyGlobal)
-		if err != nil {
-			return nil, err
-		}
+		gm.tcp4 = newPerClusterCTMap(mapTypeIPv4TCPGlobal)
+		gm.any4 = newPerClusterCTMap(mapTypeIPv4AnyGlobal)
 	}
 
 	if ipv6 {
-		gm.tcp6, err = newPerClusterCTMap(outerMapNamePrefix+perClusterTCP6OuterMapSuffix, mapTypeIPv6TCPGlobal)
-		if err != nil {
-			return nil, err
-		}
-
-		gm.any6, err = newPerClusterCTMap(outerMapNamePrefix+perClusterANY6OuterMapSuffix, mapTypeIPv6AnyGlobal)
-		if err != nil {
-			return nil, err
-		}
+		gm.tcp6 = newPerClusterCTMap(mapTypeIPv6TCPGlobal)
+		gm.any6 = newPerClusterCTMap(mapTypeIPv6AnyGlobal)
 	}
 
-	return gm, nil
-}
-
-func getInnerMapName(outerMapName string, clusterID uint32) string {
-	return outerMapName + "_" + strconv.FormatUint(uint64(clusterID), 10)
+	return &gm
 }
 
-func (gm *perClusterCTMaps) UpdateClusterCTMaps(clusterID uint32) error {
-	if err := cmtypes.ValidateClusterID(clusterID); err != nil {
-		return err
-	}
-
+func (gm *perClusterCTMaps) OpenOrCreate() (err error) {
 	gm.Lock()
 	defer gm.Unlock()
 
-	if gm.ipv4 {
-		if err := gm.tcp4.updateClusterCTMap(clusterID); err != nil {
-			return err
-		}
-
-		if err := gm.any4.updateClusterCTMap(clusterID); err != nil {
-			return err
-		}
-	}
-
-	if gm.ipv6 {
-		if err := gm.tcp6.updateClusterCTMap(clusterID); err != nil {
-			return err
-		}
+	return gm.foreach(
+		func(om *PerClusterCTMap) error { return om.OpenOrCreate() },
+	)
+}
 
-		if err := gm.any6.updateClusterCTMap(clusterID); err != nil {
-			return err
-		}
-	}
+func (gm *perClusterCTMaps) Close() (err error) {
+	gm.Lock()
+	defer gm.Unlock()
 
-	return nil
+	return gm.foreach(
+		func(om *PerClusterCTMap) error { return om.Close() },
+	)
 }
 
-func (gm *perClusterCTMaps) DeleteClusterCTMaps(clusterID uint32) error {
+func (gm *perClusterCTMaps) CreateClusterCTMaps(clusterID uint32) error {
 	if err := cmtypes.ValidateClusterID(clusterID); err != nil {
 		return err
 	}
 
 	gm.Lock()
 	defer gm.Unlock()
 
-	if gm.ipv4 {
-		if err := gm.tcp4.deleteClusterCTMap(clusterID); err != nil {
-			return err
-		}
-
-		if err := gm.any4.deleteClusterCTMap(clusterID); err != nil {
-			return err
-		}
-	}
-
-	if gm.ipv6 {
-		if err := gm.tcp6.deleteClusterCTMap(clusterID); err != nil {
-			return err
-		}
-
-		if err := gm.any6.deleteClusterCTMap(clusterID); err != nil {
-			return err
-		}
-	}
-
-	return nil
-}
-
-func (gm *perClusterCTMaps) GetClusterCTMaps(clusterID uint32) []*Map {
-	var (
-		err error
-		im  *Map
+	gm.clusterIDs.Insert(clusterID)
```

Comment:
Should we rollback this Insert when the below map creations fail?

Reply:
I'm actually unsure about that. The list of ClusterIDs is currently used only as an optimization when retrieving all maps for the GC process; in case of errors it doesn't really matter whether they are returned or not, as if the map does not exist it will just be skipped while attempting to open it.